fix: disable cal branding for user, team, organisation

dhairyashiil opened the pull request:
## What does this PR do?

- Fixes #23850 
- Fixes CAL-6414 

This PR implements the ability to hide Cal.com branding at the user, team, and organization levels. The branding visibility is determined hierarchically - if an organization has branding hidden, all teams and users under it inherit that setting.

### Key Changes

The implementation adds a `shouldHideBrandingForEvent` function that checks branding settings in this priority order:
1. Organization's `hideBranding` setting (if applicable)
2. Team's `hideBranding` setting (if applicable)  
3. Event owner's `hideBranding` setting

Email functions now use a `CalendarEventWithBranding` type to ensure branding state is explicitly passed. A `withHideBranding` helper wraps calendar events with the branding flag.

### Files Modified
- Added `hideBranding` field to `CalendarEvent` type
- Updated email-manager.ts with branding-aware function signatures
- Added repository methods for fetching branding data (TeamRepository, UserRepository, EventTypeRepository)
- Updated booking handlers, payment handlers, and cron jobs to calculate and pass branding state

## Visual Demo (For contributors especially)

User: 

https://github.com/user-attachments/assets/57190c91-7077-4d9a-b6d2-a8b27a7670ca

Team: 

https://github.com/user-attachments/assets/f7b9a053-94d2-41cf-9486-dcb1a8639ad2

Organisation:

https://github.com/user-attachments/assets/2e03ad43-7fef-4c68-9db7-ffe1f87f90c2

## Mandatory Tasks (DO NOT REMOVE)

- [ ] I have self-reviewed the code (A decent size PR without self-review might be rejected).
- [ ] I have updated the developer docs in /docs if this PR makes changes that would require a [documentation change](https://cal.com/docs). N/A - no docs changes needed.
- [ ] I confirm automated tests are in place that prove my fix is effective or that my feature works.

## How should this be tested?

1. Set up a user with `hideBranding: true` in their profile
2. Create a booking with that user as the organizer
3. Verify the confirmation email does not show Cal.com branding

4. Set up a team with `hideBranding: true`
5. Create a booking for a team event type
6. Verify the confirmation email does not show Cal.com branding

7. Set up an organization with `hideBranding: true`
8. Create bookings for users/teams under that organization
9. Verify branding is hidden for all bookings under the organization

## Human Review Checklist

- [ ] Verify `shouldHideBrandingForEvent` logic correctly handles all edge cases (null team, null owner, etc.)
- [ ] Check that all email sending paths correctly pass `hideBranding` 
- [ ] Verify type assertions in `addGuests.handler.ts` and `confirm.handler.ts` are safe
- [ ] Confirm repository queries include all necessary `hideBranding` fields

<!-- Link to Devin run: https://app.devin.ai/sessions/2876dafc0ac24155bb2ed3f2e767906b -->
<!-- Requested by: rajiv@cal.com (@Ryukemeister) -->

dhairyashiil committed:
fix: disable cal branding

dhairyashiil committed:
fix: disable branding in cancelation email

dhairyashiil committed:
fix: disable branding for confirmation booking email

dhairyashiil committed:
disable branding in seated bookings

dhairyashiil committed:
disable branding for declined bookings

dhairyashiil committed:
avoid unnecessary db call when emails disabled

dhairyashiil committed:
made hideBranding mandatory for email functions

dhairyashiil committed:
Remove unused code

dhairyashiil committed:
resolve merge conflicts

dhairyashiil committed:
Addressed coderabits comments

dhairyashiil committed:
Addressed coderabits comments

dhairyashiil committed:
Addressed coderabits comments

dhairyashiil committed:
Addressed coderabits comments

dhairyashiil committed:
no primsa outside repositories structure, better error handling, no duplicate code

dhairyashiil committed:
Resolve merge conflicts

dhairyashiil committed:
fix repository pattern violation

dhairyashiil committed:
type export and JSDoc

dhairyashiil committed:
fix: compute hideBranding dynamically in payment handlers

dhairyashiil committed:
Resolve merge conflicts

dhairyashiil committed:
fix failing round robin test

dhairyashiil committed:
fix violation of no prisma outside of repositories rule

dhairyashiil committed:
Merge remote-tracking branch 'upstream/main' into fix/disable-cal-branding-user-team-organisation

dhairyashiil committed:
update orgId passing logic and logger

dhairyashiil committed:
address cubics comments

dhairyashiil committed:
use proper types instead of any

dhairyashiil committed:
remove comments

dhairyashiil committed:
resolve merge conflicts

dhairyashiil committed:
Merge remote-tracking branch 'upstream/main' into fix/disable-cal-branding-user-team-organisation

dhairyashiil committed:
use prisma from repository

dhairyashiil committed:
address cubics comment

dhairyashiil committed:
address cubics comment

dhairyashiil committed:
Merge remote-tracking branch 'upstream/main' into fix/disable-cal-branding-user-team-organisation

Udit-takkar reviewed (commented):
@dhairyashiil type checks are failing

Udit-takkar commented:
> Type checks are passing in my local, checked multiple times:

Check CI. looks like wrong file path

Udit-takkar commented on packages/features/bookings/lib/handleCancelBooking.ts:
```diff
@@ -251,9 +252,32 @@ async function handler(input: CancelBookingInput) {
+  log.debug("Computed hideBranding", {
+    hideBranding,
+    eventTypeId: bookingToDelete.eventTypeId,
+  });
```
can we remove all these logs?

dhairyashiil replied:
removed all these types of logs (console and logger) from all files

Udit-takkar committed:
chore: trigger again

Udit-takkar commented on packages/features/profile/lib/hideBranding.ts:
```diff
@@ -179,3 +257,51 @@ export function shouldHideBrandingForUserEvent({
+  userId,
+  teamId,
+  prisma,
+}: {
```
We can do this in a follow up PR if you want. Something like 

```
import type { ILogger } from "@calcom/features/webhooks/lib/interface/infrastructure";

import type { IBrandingRepository } from "../../../profile/lib/branding/interface/repository";
import type { IBrandingService } from "../../../profile/lib/branding/interface/services";
import type { BrandingContext, BookingWithBrandingData } from "../../../profile/lib/branding/dto/types";
import {
  resolveTeamBranding,
  resolveUserBranding,
} from "../../../profile/lib/branding/resolvers/branding.resolver";

export class BrandingService implements IBrandingService {
  private readonly log: ILogger;

  constructor(
    private readonly repository: IBrandingRepository,
    logger: ILogger
  ) {
    this.log = logger.getSubLogger({ prefix: ["[BrandingService]"] });
  }

  async getForBookingId(bookingId: number): Promise<boolean> {
    const data = await this.repository.getBookingBrandingData(bookingId);
    
    if (!data) {
      this.log.warn(`Booking ${bookingId} not found`);
      return false;
    }

    return this.resolveFromBookingData(data);
  }

  async getForEventType(eventTypeId: number): Promise<boolean> {
    const data = await this.repository.getEventTypeBrandingData(eventTypeId);

    if (!data) {
      this.log.warn(`EventType ${eventTypeId} not found`);
      return false;
    }

    if (data.team) {
      return resolveTeamBranding(data.team);
    }

    if (data.user) {
      const organizationId = data.user.profiles[0]?.organizationId ?? null;
      return resolveUserBranding(data.user, organizationId);
    }

    return false;
  }

  async getForTeam(teamId: number): Promise<boolean> {
    const team = await this.repository.getTeamBrandingData(teamId);

    if (!team) {
      this.log.warn(`Team ${teamId} not found`);
      return false;
    }

    return resolveTeamBranding(team);
  }

  async getForUser(userId: number, organizationId?: number | null): Promise<boolean> {
    const user = await this.repository.getUserBrandingData(userId);

    if (!user) {
      this.log.warn(`User ${userId} not found`);
      return false;
    }

    return resolveUserBranding(user, organizationId);
  }

  resolveFromBookingData(booking: BookingWithBrandingData): boolean {
    if (booking.eventType?.team) {
      return resolveTeamBranding(booking.eventType.team);
    }

    if (booking.user) {
      const organizationId = booking.user.profiles[0]?.organizationId ?? null;
      return resolveUserBranding(booking.user, organizationId);
    }

    return false;
  }

  resolveFromContext(context: BrandingContext): boolean {
    switch (context.type) {
      case "team":
        return resolveTeamBranding(context.team);
      case "user":
        return resolveUserBranding(context.user, context.organizationId);
      case "none":
        return false;
    }
  }
}
```

Udit-takkar commented on packages/trpc/server/routers/viewer/bookings/addGuests.handler.ts:
```diff
@@ -86,6 +89,46 @@ async function getBooking(bookingId: number) {
+    booking.eventTypeId
+      ? eventTypeRepository.findOwnerHideBranding({ eventTypeId: booking.eventTypeId })
+      : Promise.resolve(null),
+  ]);
```
why don't we fetch all these inside `findByIdIncludeDestinationCalendar`?

Udit-takkar committed:
fix: type errro

dhairyashiil committed:
Merge remote-tracking branch 'upstream/main' into fix/disable-cal-branding-user-team-organisation

dhairyashiil committed:
added new method findByIdIncludeDestinationCalendarAndBranding and removed unnenessary logs

dhairyashiil commented on packages/trpc/server/routers/viewer/bookings/editLocation.handler.ts:
```diff
@@ -284,12 +287,26 @@ export async function editLocationHandler({ ctx, input }: EditLocationOptions) {
+    const eventTypeId = booking.eventTypeId;
+    let hideBranding = false;
+    if (!eventTypeId) {
+      logger.warn("Booking missing eventTypeId, defaulting hideBranding to false");
```
removed

dhairyashiil commented:
> Check CI. looks like wrong file path.
> 
> Edit: fixed

So I was importing 
`import { EventTypeRepository } from "@calcom/features/eventtypes/repositories/EventTypeRepository";`
instead of `import { EventTypeRepository } from "@calcom/features/eventtypes/repositories/eventTypeRepository";`
— and is this causing the failing type check?

dhairyashiil committed:
address cubics comments

dhairyashiil committed:
Merge remote-tracking branch 'upstream/main' into fix/disable-cal-branding-user-team-organisation

dhairyashiil committed:
Merge remote-tracking branch 'upstream/main' into fix/disable-cal-branding-user-team-organisation

dhairyashiil committed:
use correct organization resolution for both team and user events

dhairyashiil committed:
Merge remote-tracking branch 'upstream/main' into fix/disable-cal-branding-user-team-organisation

dhairyashiil committed:
use getOrgIdFromMemberOrTeamId to compute orgId in other files

dhairyashiil committed:
use helped function instead of manually computing hidebranding

dhairyashiil committed:
use select instead of include, correctly compute orgId in reminder emails, dont log sensitive info

dhairyashiil committed:
resolved merge conflicts

dhairyashiil committed:
resolved merge conflicts

dhairyashiil committed:
fix failing type check

Udit-takkar reviewed (dismissed):
LGTM

dhairyashiil committed:
Merge branch 'main' into fix/disable-cal-branding-user-team-organisation

volnei commented on packages/features/profile/lib/hideBranding.ts:
```diff
@@ -179,3 +257,51 @@ export function shouldHideBrandingForUserEvent({
+ * Convenience function that creates repositories with a PrismaClient instance
+ * This maintains backward compatibility for existing code
+ */
+export async function shouldHideBrandingForEventWithPrisma({
```
we should never use provider names for services or libraries.

volnei reviewed (changes requested):
hey @dhairyashiil ideally we should introduce Repositories for everything that touches the database. Also we should never use provider names e.g.: prisma on service or libraries.

dhairyashiil committed:
repository pattern and update method name

dhairyashiil committed:
Merge upstream-fix-temp: Complete repository pattern refactoring

- Removed WithPrisma functions from hideBranding.ts
- Created BrandingRepository for database access
- Updated all 21 call sites to use new function names
- Integrated with new organizationId helper functions
- Resolved merge conflicts with email-manager import changes

dhairyashiil committed:
update one missed file change

volnei reviewed (dismissed):
@dhairyashiil we still have this `shouldHideBrandingForEventWithPrisma`.

dhairyashiil committed:
Merge remote-tracking branch 'upstream/main' into fix/disable-cal-branding-user-team-organisation

dhairyashiil committed:
Merge upstream/main into fix/disable-cal-branding-user-team-organisation

volnei commented:
@cubic-dev-ai can you analise possible regressions, missing tests and risks of this PR?

dhairyashiil committed:
fix: use dynamic hideBranding flag instead of hardcoded true in email handlers

- Replace all withHideBranding(..., true) calls with evt.hideBranding ?? false
- Ensures branding visibility respects user/team/organization subscription settings
- Fixes regression where all emails would hide branding regardless of settings
- Updated BookingEmailSmsHandler.ts (5 occurrences) and videoClient.ts (2 occurrences)

dhairyashiil committed:
perf: remove unused eventType.parent select in connectAndJoin handler

- Remove unnecessary parent.teamId select that was never used in the handler
- Reduces database query complexity by eliminating extra relation lookup
- Keeps only required selects: parentId and team.parent.hideBranding
- Follows 'only select what you need' best practice for optimal performance

dhairyashiil committed:
refactor: replace include with select in findByIdIncludeDestinationCalendarAndBranding

- Convert attendees from include to select (only id, email, name, timeZone, locale)
- Convert references from include to select (only video call related fields)
- Add explicit booking field selects to avoid pulling all columns
- Reduces PII exposure and improves query performance
- Follows repository pattern policy of 'select what you need'

dhairyashiil committed:
perf: remove unused eventType.parent select in handleDeleteCredential

- Remove unnecessary parent.teamId select that was never used in the handler
- Only team.parentId and team.parent.hideBranding are actually needed
- Reduces Prisma payload size and improves query performance
- Follows 'only select what you need' best practice

dhairyashiil committed:
Merge remote-tracking branch 'upstream/main' into fix/disable-cal-branding-user-team-organisation

dhairyashiil committed:
fix: add missing userPrimaryEmail field to BookingRepository select

- Add userPrimaryEmail: true to findByIdIncludeDestinationCalendarAndBranding
- Prevents regression where organizer email would fall back to user record
- Ensures existing bookings preserve their stored organizer address

dhairyashiil committed:
resolve merge conflicts

dhairyashiil committed:
Merge upstream/main into fix/disable-cal-branding-user-team-organisation

Resolved conflicts by keeping both PR's hideBranding logic and main's new features:
- email-manager.ts: Keep PR's CalendarEventWithBranding type
- BookingEmailSmsHandler.ts: Merge imports and hideBranding usage
- getBookingToDelete.ts: Keep both hideBranding and profiles
- handleCancelBooking.ts: Keep both imports and hideBranding + creditCheckFn
- handleConfirmation.ts: Keep hideBranding calculation + tracing from main
- handleNoShowFee.ts: Add both hideBranding and profiles to user type
- RegularBookingService.ts: Keep both withBranding and withOrganization
- BookingRepository.ts: Merge imports and keep both methods
- roundRobinManualReassignment.ts: Keep PR's email functions with hideBranding + creditCheckFn
- roundRobinReassignment.ts: Keep PR's email functions with hideBranding
- WorkflowService.test.ts: Keep both mock sets
- confirm.handler.ts: Keep both imports
- util.ts: Add both hideBranding and profiles to type
- eventTypeRepository.ts: Keep both findOwnerHideBranding and new methods from main
- handleDeleteCredential.ts: Keep both hideBranding and profiles

dhairyashiil committed:
fix failing tests and type check

dhairyashiil committed:
fix failing tests

dhairyashiil committed:
fix failing tests

Ryukemeister committed:
Merge branch 'main' into fix/disable-cal-branding-user-team-organisation

Ryukemeister committed:
chore: update locales

Ryukemeister committed:
fix: hide branding should be respected when sending verify email subject

Ryukemeister committed:
Merge branch 'main' into fix/disable-cal-branding-user-team-organisation

Ryukemeister committed:
fix: bad imports

Ryukemeister committed:
chore: implement cubic feedback

Ryukemeister committed:
Merge branch 'main' into fix/disable-cal-branding-user-team-organisation

keithwillcode committed:
Merge branch 'main' into fix/disable-cal-branding-user-team-organisation

Ryukemeister committed:
fixup: implement correct logic for booker email verification for hide branding

dhairyashiil committed:
resolve merge conflicts

Ryukemeister committed:
Merge branch 'main' into fix/disable-cal-branding-user-team-organisation

Ryukemeister reviewed (dismissed):
nice one @dhairyashiil! one small feedback tho, it would have been much safer and easier to review if had multiple PRs for this with, that way we won't have so many lines changes and file changes in a single PR so it would be easier to understand whats going on. other than good, all good!

hbjORbj reviewed (changes requested):
Unit tests failing

hbjORbj commented on packages/features/profile/lib/hideBranding.ts:
```diff
@@ -2,35 +2,60 @@ import { TeamRepository } from "@calcom/features/ee/teams/repositories/TeamRepos
+// Internal type aliases for backward compatibility
+type Team = TeamWithBranding;
+type UserWithoutProfile = UserWithBranding;
+type UserWithProfile = UserWithProfileAndBranding;
```
not optimal, just use the new names

hbjORbj commented on packages/emails/email-manager.ts:
```diff
@@ -46,6 +46,12 @@ import OrganizerRequestedToRescheduleEmail from "./templates/organizer-requested
+
+export function withHideBranding(calEvent: CalendarEvent, explicit?: boolean): CalendarEventWithBranding {
+  return { ...calEvent, hideBranding: explicit ?? calEvent.hideBranding ?? false };
+}
```
any particular reason for adding this in /packages/emails?

Ryukemeister replied:
makes sense, `withHideBranding` looks a bit odd here. shifted this to `hideBranding` instead.

devin-ai-integration[bot] committed:
Merge upstream/main into fix/disable-cal-branding-user-team-organisation

Resolved merge conflicts in:
- packages/app-store/_utils/payments/handlePaymentSuccess.ts
- packages/features/bookings/lib/handleCancelBooking.ts
- packages/features/bookings/lib/handleConfirmation.ts
- packages/features/bookings/lib/handleSeats/reschedule/attendee/attendeeRescheduleSeatedBooking.ts
- packages/features/bookings/lib/handleSeats/reschedule/owner/moveSeatedBookingToNewTimeSlot.ts
- packages/features/ee/payments/api/webhook.ts
- packages/features/eventtypes/repositories/eventTypeRepository.ts
- packages/trpc/server/routers/viewer/bookings/addGuests.handler.ts
- packages/trpc/server/routers/viewer/bookings/confirm.handler.ts

Preserved PR's hideBranding feature implementation while incorporating upstream changes.

Co-Authored-By: unknown <>

Ryukemeister committed:
chore: implement PR feedback

Ryukemeister committed:
Merge branch 'main' into fix/disable-cal-branding-user-team-organisation

Ryukemeister committed:
fix: type check

Ryukemeister committed:
fix: unit tests

Ryukemeister committed:
chore: shift withHideBranding fn from email-manager to hideBranding

Ryukemeister committed:
chore: implement cubic feedback

Ryukemeister committed:
Merge branch 'main' into fix/disable-cal-branding-user-team-organisation

hbjORbj reviewed (dismissed):
Build and unit tests failing. Putting this back to draft.

Ryukemeister committed:
Merge branch 'main' into fix/disable-cal-branding-user-team-organisation

hbjORbj reviewed (dismissed):
I see lots of formatting changes in this PR, making the review quite difficult. I suggest using Devin or some AI to revert the formatting changes so it's easy to spot the actual changes this PR brings

devin-ai-integration[bot] committed:
Merge upstream/main into fix/disable-cal-branding-user-team-organisation

Co-Authored-By: rajiv@cal.com <sahalrajiv6900@gmail.com>

devin-ai-integration[bot] committed:
fix: revert formatting changes to improve PR readability

Co-Authored-By: rajiv@cal.com <sahalrajiv6900@gmail.com>

devin-ai-integration[bot] committed:
refactor: remove unused findByIdIncludeDestinationCalendar method

Co-Authored-By: rajiv@cal.com <sahalrajiv6900@gmail.com>

Ryukemeister commented on packages/trpc/server/routers/viewer/bookings/addGuests.handler.ts:
```diff
@@ -103,7 +107,7 @@ export const addGuestsHandler = async ({
 
 async function getBooking(bookingId: number) {
   const bookingRepository = new BookingRepository(prisma);
-  const booking = await bookingRepository.findByIdIncludeDestinationCalendar(bookingId);
```
yeah I think so, doesnt seem to be used anywhere

Ryukemeister committed:
chore: implement cubic feedback

Ryukemeister committed:
Merge branch 'main' into fix/disable-cal-branding-user-team-organisation

Ryukemeister committed:
Merge branch 'main' into fix/disable-cal-branding-user-team-organisation

Ryukemeister reviewed (commented):
hey @dhairyashiil, as discussed we should move this into multiple smaller PRs. this is insanely large and has a higher chance of breaking. i'll create follow up PRs for this. for now we can close this, thanks!
